### DISCUSSION TITLE
Filtered fix run using `ESLint`

### DISCUSSION THREAD

**Discussion by bradzacher** [eslint/eslint#16014] [General]
I'm working on trying to get the Meta codebases into a place wherein we can upgrade to ESLint v8.
This, of course, involves updating our infra to use `ESLint` instead of `CLIEngine`.

For the most part this is fine - most of the infra is just doing a bog standard lint run.

The one thing that's complicated is our IDE extension which does autofix-on-save using an allowlist. Put another way - when you save the file we don't allow every single rule's fixer to be applied; instead we only allow a smaller set of hand-picked rules to be fixed.
This provides a better DevX by ensuring only high quality fixers that can be applied to "unfinished" code are run.

I'm running into a bit of a dead end in regards implementing this using the `ESLint` class.

<details>
<summary>
  With <code>CLIEngine</code> we achieved this by pairing it with a <code>Linter</code>.
</summary>

```ts
function fixAll(
  engine: CLIEngine,
  text: string,
  path: string,
  allowedRules: string[]
): string | null {
  const config = engine.getConfigForFile(path);
  const rulesMap = engine.getRules();
  
  // determine the new config
  const oldRulesForFile = config.rules ?? {};
  const newRulesForFile = {};
  for (const ruleId of allowedRules) {
    if (ruleId in oldRulesForFile) {
      newRulesForFile[ruleId] = oldRulesForFile[ruleId];
    }
  }

  // prep a linter
  const linter = new Linter();
  // ESLint has already resolved config.parser to the absolute path for us
  linter.defineParser(config.parser, require(config.parser));
  for (const ruleId of allowedRules) {
    const rule = rulesMap.get(ruleId);
    if (rule) {
      linter.defineRule(ruleId, rule);
    }
  }

  // apply the fixers
  const result = linter.verifyAndFix(text, {...config, rules: newRulesForFile}, path);
  return result.output;
}
```

</details>

However it looks like I can't do this with `ESLint` because one of the crucial bits of this doesn't exist - specifically `getRules`.
`getRulesMetaForResults` *almost* does what what we want, except it only returns the **metadata**, not the full rule module.
This means there's no way to ask the instance for the pre-loaded rule modules.

I've tried exploring a few directions
1) configure a `Linter` instance by manually loading plugins
    - given that plugins can be loaded relative to configs - I don't think this is ultimately feasible to do?
2) construct a brand new `ESLint` instance by getting and then "reducing" the config.
    - the issue with this is that I can't get the original options that were passed to the `ESLint` constructor - so it's hard to keep things consistent and correct based on the rest of our infra.
3) use `@eslint/eslintrc`'s utils the same way `CLIEngine` does
    - that code is *supposed* to be internal, so this isn't a great idea.
4) filter the results and manually apply the fixes to the file using the same logic ESLint does
    - this is essentially what [vscode-eslint](https://github.com/microsoft/vscode-eslint) does, but it's a lot of code to maintain / reimplement.

None of these are fantastic, sadly.
So I'm wondering - am I missing something here, or am I correct in saying that it's not possible using just `ESLint`?
If I am right - is this something we can consider adding to `ESLint`? I'd be happy to write up an RFC!

**Comment by mdjermanovic** [comment]
`fix` option of the [ESLint constructor](https://eslint.org/docs/developer-guide/nodejs-api#-new-eslintoptions) can accept a function that will be called for each reported lint message and should return `true` for those whose fixes should be applied, `false` for those that shouldn't be applied. Lint messages have `ruleId` property that could be checked against the allowlist, in the function. Can that be helpful for your use case?

**Reply by bradzacher** [reply]
"sort of" for our usecase you have to assume we have an instance already and we cannot create a new one.

This is where it gets difficult. There's two easy ways to solve this:
1) construct a new `ESLint` instance with a config that's been filtered. This would be preferred as it means you only run the mint rules being fixed.
2) construct a new `ESLint` instance and pass a `fix` function. This would work but it would be a little wasteful.

Note that we can't *always* run the filtered config because we want to show all lint rules and allow manual inline fixers for them - we just want to do a filtered run on save.

I guess what I'm looking for is either a method that lets you change the config after construction (eg let me set the override config whenever), or a method that gives me all the rule modules so I can do dynamic stuff with a `Linter`.
